node/beacon/src/authority.rs

```diff
@@ -251,51 +258,68 @@ impl Authority {
+                }
                 self.compute_threshold_accepted(
                     &CryptoHash::default(),
-                    proposals.collect(),
+                    proposals,
                     rollovers,
                 )
             };
-            self.thresholds.insert(epoch, new_threshold);
+            storage.set_threshold(epoch, new_threshold);
             let slots: Vec<_> = self.epoch_to_slots(epoch).collect();
-            self.accepted_authorities
-                .extend(slots.iter().cloned().zip(accepted_authorities.drain(..)));
+            storage.extend_accepted_authorities(
+                slots.iter().cloned().zip(accepted_authorities.drain(..)).collect()
+            );
         }
     }
 
     /// Record proposals and participation from the given block.
     pub fn process_block_header(&mut self, header: &SignedBeaconBlockHeader) {
         // Skip genesis block or if this block was already recorded.
         let slot = header.body.index;
-        if slot > 0 && !self.proposals.contains_key(&slot) {
-            self.proposals.insert(slot, header.body.authority_proposal.to_vec());
-            self.participation.insert(slot, header.signature.authority_mask.to_vec());
+        if slot > 0 && self.storage.write().expect(POISONED_LOCK_ERR).get_proposal(slot).is_none() {
+            let (all_slots_processed, epoch) = {
+                let mut storage = self.storage.write().expect(POISONED_LOCK_ERR);
+                storage.set_proposal(slot, header.body.authority_proposal.to_vec());
+                storage.set_participation(slot, header.signature.authority_mask.to_vec());
 
-            // Update the tracker of processed slots.
-            let epoch = self.slot_to_epoch(slot);
-            let all_slots_processed = {
-                let processed_slots =
-                    self.processed_blocks.entry(epoch).or_insert_with(HashSet::new);
+                // Update the tracker of processed slots.
+                let epoch = self.slot_to_epoch(slot);
+                let mut processed_slots = 
+                    if let Some(slots) = storage.get_processed_blocks(epoch) {
+                        slots.clone()
+                    } else {
+                        HashSet::new()
+                    };
                 processed_slots.insert(slot);
-                processed_slots.len() == self.authority_config.epoch_length as usize
+                let len = processed_slots.len();
+                storage.set_processed_blocks(epoch, processed_slots);
+                (len == self.authority_config.epoch_length as usize, epoch)
             };
+            
             // Check if we have processed all slots from the given epoch.
             if all_slots_processed {
                 // Compute accepted authorities for epoch+2.
                 self.compute_accepted_authorities(epoch + 2);
+                self.storage.write().expect(POISONED_LOCK_ERR).prune_authority_storage(
```

Comment:
So pruning currently heavily relies on the assumption that once we prune epoch X, we will never need to know authorities for this epoch, because if we do (for whatever reason) authority has no mechanism to ask client to re-feed it the old headers of blocks in epoch X. An attempt to get pruned authorities, however should then either panic or return `Err`. If we return `Err` then the client code that uses it will need an additional layer of logic that recovers from it. Pruning seems to be making our code more complex at the moment. Also, it is not decided yet whether we should be pruning blocks/headers, if we do not then it does not make much sense to prune authorities because they are negligibly smaller than blocks/headers.

I suggest we do to things right now:
* Add code to the client that recovers from the `Err` that we should return when the pruned authorities are requested;
* Have a separate issue for other pruning of the storage, like blocks, headers, transactions, etc.

Reply:
How does client recover from the error we return?

Reply:
It should depends on the where in the clients code we request authorities. This would require an investigation on your part, go through every part of client that calls authority and make sure that if it requests a pruned authority then it does not just crash and burn.

Reply:
Fixed